pip install error (x86_64-linux-gnu-gcc) on ubuntu16.04

I am getting error when "Pip3 install tesserocr" on my ubuntu16.04 machine/


> pip3 install tesseractocr
> Collecting tesseractocr
>   Could not find a version that satisfies the requirement tesseractocr (from versions: )
> No matching distribution found for tesseractocr
> root@70d41ee0a519:/tmp# pip3 install tesserocr
> Collecting tesserocr
>   Downloading https://files.pythonhosted.org/packages/f8/6d/4e81e041f33a4419e59edcb1dbdf3c56e9393f60f5ef531381bd67a1339b/tesserocr-2.3.1.tar.gz (54kB)
>     100% |################################| 61kB 3.7MB/s
> Building wheels for collected packages: tesserocr
>   Running setup.py bdist_wheel for tesserocr ... error
>   Complete output from command /usr/bin/python3.4 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-ocheer6s/tesserocr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-3he8anm_ --python-tag cp34:
>   Supporting tesseract v3.04.01
>   Configs from pkg-config: {'libraries': ['lept', 'tesseract'], 'include_dirs': ['/usr/include'], 'cython_compile_time_env': {'TESSERACT_VERSION': 197633}}
>   running bdist_wheel
>   running build
>   running build_ext
>   building 'tesserocr' extension
>   creating build
>   creating build/temp.linux-x86_64-3.4
>   x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include -I/usr/include/python3.4m -c tesserocr.cpp -o build/temp.linux-x86_64-3.4/tesserocr.o
>   cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
>   tesserocr.cpp:38:20: fatal error: Python.h: No such file or directory
>   compilation terminated.
>   error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
> 
>   ----------------------------------------
>   Failed building wheel for tesserocr
>   Running setup.py clean for tesserocr
> Failed to build tesserocr
> Installing collected packages: tesserocr
>   Running setup.py install for tesserocr ... error
>     Complete output from command /usr/bin/python3.4 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-ocheer6s/tesserocr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-avohldt9/install-record.txt --single-version-externally-managed --compile:
>     Supporting tesseract v3.04.01
>     Configs from pkg-config: {'cython_compile_time_env': {'TESSERACT_VERSION': 197633}, 'libraries': ['tesseract', 'lept'], 'include_dirs': ['/usr/include']}
>     running install
>     running build
>     running build_ext
>     building 'tesserocr' extension
>     creating build
>     creating build/temp.linux-x86_64-3.4
>     x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include -I/usr/include/python3.4m -c tesserocr.cpp -o build/temp.linux-x86_64-3.4/tesserocr.o
>     cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
>     tesserocr.cpp:38:20: fatal error: Python.h: No such file or directory
>     compilation terminated.
>     error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
> 
>     ----------------------------------------
> Command "/usr/bin/python3.4 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-ocheer6s/tesserocr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-avohldt9/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-ocheer6s/tesserocr/


Here is my tesseract version

> tesseract 3.04.01
>  leptonica-1.73
>   libgif 5.1.2 : libjpeg 8d (libjpeg-turbo 1.4.2) : libpng 1.2.54 : libtiff 4.0.6 : zlib 1.2.8 : libwebp 0.4.4 : libopenjp2 2.1.2


Any hints?

Thanks

--- a/tesseract.pxd
+++ b/tesseract.pxd
@@ -36,6 +36,7 @@ cdef extern from "leptonica/allheaders.h" nogil:
     int pixWriteMemJpeg(unsigned char **, size_t *, Pix *, int, int)
     int pixWriteMem(unsigned char **, size_t *, Pix *, int)
     void pixDestroy(Pix **)
+    void ptaDestroy(Pta **)
     int setMsgSeverity(int)
     void pixaDestroy(Pixa **)
     void boxaDestroy(Boxa **)

--- a/tesserocr.pyx
+++ b/tesserocr.pyx
@@ -612,12 +612,15 @@ cdef class PyPageIterator:
         if pta == NULL:
             return None
         try:
-            return zip((x for x in pta.x[:pta.n]), (y for y in pta.y[:pta.n]))
+            return list(zip((x for x in pta.x[:pta.n]), (y for y in pta.y[:pta.n])))
         finally:
-            free(pta)
+            ptaDestroy(pta)
 
     def GetBinaryImage(self, PageIteratorLevel level):
         """Return a binary image of the current object at the given level.
+
+        The image is masked along the polygon outline of the current block, as given
+        by :meth:`BlockPolygon`. (Pixels outside the mask will be white.)
 
         The position and size match the return from :meth:`BoundingBoxInternal`, and so
         this could be upscaled with respect to the original input image.
@@ -639,6 +642,9 @@ cdef class PyPageIterator:
     def GetImage(self, PageIteratorLevel level, int padding, original_image):
         """Return an image of the current object at the given level in greyscale
         if available in the input.
+
+        The image is masked along the polygon outline of the current block, as given
+        by :meth:`BlockPolygon`. (Pixels outside the mask will be white.)
 
         To guarantee a binary image use :meth:`BinaryImage`.
 

--- a/tesserocr.pyx
+++ b/tesserocr.pyx
@@ -614,7 +614,7 @@ cdef class PyPageIterator:
         try:
             return list(zip((x for x in pta.x[:pta.n]), (y for y in pta.y[:pta.n])))
         finally:
-            ptaDestroy(pta)
+            ptaDestroy(&pta)
 
     def GetBinaryImage(self, PageIteratorLevel level):
         """Return a binary image of the current object at the given level.
@@ -2344,6 +2344,7 @@ cdef class PyTessBaseAPI:
                   rotation to be applied to the page for the text to be upright and readable.
                 - oconfidence: Orientation confidence.
                 - script: Index of the script with the highest score for this orientation.
+                  (This is _not_ the index of :meth:`get_languages`, which is in alphabetical order.)
                 - sconfidence: script confidence.
         """
         cdef OSResults results

--- a/tests/test_api.py
+++ b/tests/test_api.py
@@ -266,6 +266,51 @@ class TestTessBaseApi(unittest.TestCase):
         # Test if empty
         self.assertFalse(result)
 
+    def test_layout_getcomponents(self):
+        self._api.Init()
+        self._api.SetImageFile(self._image_file)
+        result = self._api.GetComponentImages(tesserocr.RIL.BLOCK, True)
+        # Test if not empty
+        self.assertTrue(result)
+        _, xywh, _, _ = result[0] # bbox of largest
+        self.assertIn('w', xywh)
+        self.assertIn('h', xywh)
+        area = xywh['w'] * xywh['h']
+        # Test if the largest block is quite large
+        self.assertGreater(area, 400000)
+
+    def test_layout_boundingbox(self):
+        self._api.Init()
+        self._api.SetImageFile(self._image_file)
+        layout = self._api.AnalyseLayout()
+        # Test if not empty
+        self.assertTrue(layout)
+        self.assertFalse(layout.Empty(tesserocr.RIL.BLOCK))
+        result = layout.BoundingBox(tesserocr.RIL.BLOCK) # bbox of largest
+        self.assertIsNot(result, None)
+        x0, y0, x1, y1 = result
+        area = (x1 - x0) * (y1 - y0)
+        # Test if the largest block is quite large
+        self.assertGreater(area, 400000)
+
+    def test_layout_blockpolygon(self):
+        self._api.Init()
+        self._api.SetImageFile(self._image_file)
+        layout = self._api.AnalyseLayout()
+        # Test if not empty
+        self.assertTrue(layout)
+        self.assertFalse(layout.Empty(tesserocr.RIL.BLOCK))
+        result = layout.BlockPolygon() # polygon of largest
+        # Test if not empty
+        self.assertIsNot(result, None)
+        # Test there are at least 4 contour points
+        self.assertGreaterEqual(len(result), 4)
+        xs, ys = zip(*result)
+        x0, y0, x1, y1 = min(xs), min(ys), max(xs), max(ys)
+        area = (x1 - x0) * (y1 - y0)
+        # Test if the largest block is quite large
+        self.assertGreater(area, 400000)
+
     def test_recognize(self):
         """Test Recognize with and without timeout."""
         self._api.SetImageFile(self._image_file)

--- a/tests/test_api.py
+++ b/tests/test_api.py
@@ -225,7 +225,10 @@ class TestTessBaseApi(unittest.TestCase):
         orientation = self._api.DetectOS()
         all(self.assertIn(k, orientation) for k in ['sconfidence', 'oconfidence', 'script', 'orientation'])
         self.assertEqual(orientation['orientation'], 0)
-        self.assertEqual(orientation['script'], 1)
+        languages = tesserocr.get_languages()[1] # this is sorted alphabetically!
+        self.assertLess(orientation['script'], len(languages))
+        script_name = languages[orientation['script']] # therefore does not work
+        #self.assertEqual(script_name, 'Latin') # cannot test: not reliable
         if _TESSERACT_VERSION >= 0x3999800:
             orientation = self._api.DetectOrientationScript()
             all(self.assertIn(k, orientation) for k in ['orient_deg', 'orient_conf', 'script_name', 'script_conf'])

